### PR TITLE
Add SQLAlchemy ORM models

### LINKED ISSUE #120: Add a beamlinesetup model
> Preferred solution: add a beamlinesetup model and make it reachable via the data collection model :slightly_smiling_face:

_Originally posted by @stufisher in https://github.com/DiamondLightSource/ispyb-database/issues/64#issuecomment-775007407_

### PR DESCRIPTION
**Pull request by rjgildea**
* New module `ispyb.sqlalchemy`:
  * Auto-generated ORM models via [sqlacodegen](https://pypi.org/project/sqlacodegen/) into `ispyb/sqlalchemy/_auto_db_schema.py`
  * Models are imported to and accessed via `ispyb.sqlalchemy`
  * Any modifications to models (e.g. injecting additional relationships between models) should be made in `ipsyb/sqlalchemy/__init__.py`
* New `ispyb.last_data_collections_on` command-line interface
* Update to ispyb-database v1.18.1 (requires ispyb-api#122)

Presumably this will be a major release, i.e. 6.0.0.

### CODE REVIEW

**Comment by rjgildea** [comment]
@Anthchirp do you want to deprecate `ispyb.model` as part of this PR? What deprecation schedule would you prefer, and what would be the best mechanism? I guess a module-level import warning would probably suffice?

**Comment by Anthchirp** [comment]
Will do that separately after this is merged

File: .azure-pipelines/ci.yml [hunk 1]
```diff
@@ -59,6 +59,7 @@ steps:
     done
     mysql -e "CREATE USER ispyb_api@'%' IDENTIFIED BY 'password_1234'; GRANT ispyb_processing to ispyb_api@'%'; GRANT ispyb_import to ispyb_api@'%'; SET DEFAULT ROLE ispyb_processing FOR ispyb_api@'%';"
     mysql -e "CREATE USER ispyb_api_future@'%' IDENTIFIED BY 'password_4321'; GRANT SELECT ON ispybtest.* to ispyb_api_future@'%';"
+    mysql -e "CREATE USER ispyb_api_sqlalchemy@'%' IDENTIFIED BY 'password_5678'; GRANT SELECT ON ispybtest.* to ispyb_api_sqlalchemy@'%'; GRANT INSERT ON ispybtest.* to ispyb_api_sqlalchemy@'%'; GRANT UPDATE ON ispybtest.* to ispyb_api_sqlalchemy@'%';"
```
Comment (Anthchirp) [inline .azure-pipelines/ci.yml:62]: Should this be read-only?

File: HISTORY.rst [hunk 2]
```diff
@@ -4,6 +4,12 @@ History
 
 Unreleased / master
 -------------------
+
+6.0.0 (XXXX-XX-XX)
+------------------
+* New ``ispyb.sqlalchemy`` module containing `SQLAlchemy <https://www.sqlalchemy.org>`_ ORM models to interface with ISPyB
+* New ``ispyb.last_data_collections_on`` command-line interface
```
Comment (Anthchirp) [inline HISTORY.rst]: This should probably say what it does.

File: ispyb/__init__.py [hunk 3]
```diff
@@ -47,6 +47,24 @@ def open(configuration_file):
     return conn
 
 
+def sqlalchemy_session(configuration_file):
```
Comment (Anthchirp) [inline ispyb/__init__.py]: I'm wondering whether this should be a `def session(...)` in `ispyb/sqlalchemy/__init__.py`.
Further, this should probably take `configuration=None` and then
```python
configuration = configuration or os.getenv("ISPYB_CREDENTIALS")
```
but appreciate opinions may vary on both accounts

File: ispyb/cli/last_data_collections_on.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+            f"{row.GridInfo.steps_x:.0f}x{row.GridInfo.steps_y:.0f}"
+            if row.GridInfo
+            else ""
+        )
+        grid = f", {grid_size:>5} grid" if grid_size else ""
+        print(
+            f"{start_time} {bl_name:8} {dcid:8} {visit:<11} {n_images:4} images{grid}   {template}"
+        )
+        if link:
+            print(" " * 52 + f"https://ispyb.diamond.ac.uk/dc/visit/{visit}/id/{dcid}")
```
Comment (Anthchirp) [inline ispyb/cli/last_data_collections_on.py]: Is there a place in the database that determines the base URL? I'm just thinking that hardcoding this isn't very portable.

**Review by Anthchirp — APPROVED** [review]
My main remaining concern is that something should be added [to the documentation](https://ispyb.readthedocs.io/en/latest/usage.html) about this.

Otherwise this looks like a very worthwhile change, especially because it is so much more powerful than our custom-built models, while at the same time requiring a lot less manual labour to maintain it.

File: .azure-pipelines/ci.yml (hunk 1 above)
Comment (rjgildea) [reply .azure-pipelines/ci.yml:62]: This is required to insert data for testing the ProcessingJob* tables - yes, I could have used the existing stored procedures, however I thought it useful/valuable to test the sqlalchemy interface for inserting data as well as reading. The above permissions are only relevant in the context of the CI job anyway, and have no impact on outside behaviour.

https://github.com/DiamondLightSource/ispyb-api/pull/123/files#diff-3eee5431b0f1a98627dd78c8f328216c7cef8fc5e6f9bc9a003c983583c94ee4R75-R96

**Comment by rjgildea** [comment]
> My main remaining concern is that something should be added [to the documentation](https://ispyb.readthedocs.io/en/latest/usage.html) about this.

I agree that we should have some documentation regarding the new sqlalchemy interface, however I didn't want to invest significant effort in documentation before we were sure we were happy with the direction. Perhaps it is best if I add a separate issue to add documentation, then this PR doesn't need to be kept waiting on documentation?

**Comment by Anthchirp** [comment]
Works for me 👍

**Comment by Anthchirp** [comment]
Do we need any more reviews on this or can this go in? @KarlLevik

File: ispyb/cli/last_data_collections_on.py (hunk 4 above)
Comment (rjgildea) [reply ispyb/cli/last_data_collections_on.py]: It appears not, however you can now override the base SynchWeb URL via the command line, e.g.:

`ispyb.last_data_collections_on i03 --link --synchweb-url=https://wls.ac.uk`

https://github.com/DiamondLightSource/ispyb-api/pull/123/commits/ac652f51e14819b1922a2c26394eaa4bc72829c2

Comment (Anthchirp) [reply ispyb/cli/last_data_collections_on.py]: Seen it, thank you. It was actually just literally meant as a question to @KarlLevik on how we handle... localisation (for lack of a better term) in ISPyB. I could see that to be useful in more than just this one context.

**Pull request merged**